### PR TITLE
Convert assert -> cast.

### LINKED ISSUE #55868: Using `__torch_function__` with RNN module
## To Reproduce

1. Defining new Tensor following the logic described here: https://pytorch.org/docs/stable/notes/extending.html#extending-torch
2. Note that the new tensor is not subclasses `torch.Tensor`. 
3. define new function for `torch.rnn_tanh`.
4. When trying to use RNN model with the new tensor, hitting the following error:
``` 
File "/common/pytorch1.8/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/common/pytorch1.8/lib/python3.7/site-packages/torch/nn/modules/rnn.py", line 244, in forward
    assert isinstance(input, Tensor)
AssertionError
```
basically, we can't mix RNN module with the `__torch_function__` feature, if the new Tensor is not subclassing `torch.Tensor`

example of simplified version of my code:
```
HANDLED_FUNCTIONS = {}
class MyTensor(object):
    def __init__(self, torch_tensor):
         self.torch_tensor = torch_tensor
   
    def __torch_function__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        if func not in HANDLED_FUNCTIONS or not all(
            issubclass(t, (torch.Tensor, ScalarTensor))
            for t in types
        ):
            return NotImplemented
        return HANDLED_FUNCTIONS[func](*args, **kwargs)

import functools
def implements(torch_function):
    """Register a torch function override for ScalarTensor"""
    @functools.wraps(torch_function)
    def decorator(func):
        HANDLED_FUNCTIONS[torch_function] = func
        return func
    return decorator

@implements(torch.rnn_tanh)
def rnn_tanh(input, hx, weights, bias, num_layers,
             dropout_p, train, bidirectional, batch_first):
    return my_computation(...)

model = RNN(10, 20, 2)
input = MyTensor(torch.randn(1, 5, 10))
model(input)
```

## Expected behavior

To be able to mix the above behaviors when I removed the assertions in the below lines it works:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L244
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L250

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 PyTorch version: 1.8.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 11.0.0
CMake version: version 3.19.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[pip3] torch==1.8.1
[conda] Could not collect

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @hameerabbasi @rgommers @peterbell10 @malfet @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #55868.

### CODE REVIEW

File: test/test_overrides.py [hunk 1]
```diff
@@ -978,5 +978,17 @@ def test_iterator(self):
...
 
+
+class TestRNN(TestCase):
+    # Regression test for gh-55868
+    def test_rnn(self):
+        try:
+            model = torch.nn.RNN(10, 20, 2)
+            input = Wrapper(torch.randn(1, 5, 10))
+            model(input)
+        except Exception as e:
```
Comment (mrzzd) [inline test/test_overrides.py:989]: do you expect some exception raises? If yes, check explicitly for the exception; otherwise why we need try-except?

File: torch/distributed/pipeline/sync/batchnorm.py [hunk 2]
```diff
@@ -151,8 +151,8 @@ def convert_deferred_batch_norm(cls, module: TModule, chunks: int = 1) -> TModul
             if module.affine:
                 module_output.register_parameter("weight", module.weight)
                 module_output.register_parameter("bias", module.bias)
-            assert isinstance(module.running_mean, Tensor)
-            assert isinstance(module.running_var, Tensor)
+            module.running_mean = cast(module.running_mean, Tensor)
+            module.running_var = cast(module.running_var, Tensor)
```
Comment (KatiaSN602) [inline torch/distributed/pipeline/sync/batchnorm.py:155]: Just curious. Why not also change this:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L718-L719

File: test/test_overrides.py (hunk 1 above)
Comment (hameerabbasi) [reply test/test_overrides.py:989]: This can either pass or fail here, but this test needs re-writing, I've marked the PR as draft for now.

**Review by rgommers — APPROVED** [review]
This LGTM, thanks @hameerabbasi. Removal of all these asserts is quite nice.

`LSTM` and GRU` are not currently broken with tensor-like's, but adding them to this regression test would be good. If needed can be done afterwards though; this PR should go in for 1.9.0

**Comment by rgommers** [comment]
@walterddr or @mruberry could we land this before the branch cut? It fixes a fairly annoying regression.

**Review by walterddr — APPROVED** [review]
looks good to me.
